### PR TITLE
Introduce method to allow modifying Paparazzi config.

### LINKED ISSUE #194: `Paparazzi.snapshot()` does not load resources of its `deviceConfig` parameter
To reproduce:
 1. Check out [this commit](https://github.com/larryng/paparazzi/commit/d77b1a674b67673606b287050b97e41bdec40628)
 2. Record sample screenshots: `./gradlew :sample:recPapaDeb` 
 3. Note how background in `app.cash.paparazzi.sample_LaunchViewTest_mix_nexus7.png` is not red `#f00` as it should be.
 4. Swap lines 29 and 30 to set the starting config to `NEXUS_7` and re-run the tests
 5. Note how the `nexus7` background is red but the `pixel3` screenshot is now also red.

### PR DESCRIPTION
**Pull request by luis-cortes**
Resolves #194

The call to `Paparazzi.inflate` was using resources based on the DeviceConfig passed into Paparazzi's constructor and the resources will not be updated until a new SessionParams is built with the updated DeviceConfig.

This change moves several parameters from the snapshot and gif methods over to a new method `Paparazzi.unsafeConfig` which ensures that a new session is created with the updated config.
If users wish to test layout resource files for different device sizes they will be required to inflate a second time after updating Paparazzi's config.

Ex:
```
  @Test
  fun mix() {
    var launch = paparazzi.inflate<LinearLayout>(R.layout.launch)
    paparazzi.snapshot(launch, "pixel3")
    paparazzi.unsafeUpdateConfig(deviceConfig = NEXUS_7)
    launch = paparazzi.inflate(R.layout.launch)
    paparazzi.snapshot(launch, "nexus7")
  }
}
```

### CODE REVIEW

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt [hunk 1]
```diff
@@ -229,50 +214,55 @@ class Paparazzi @JvmOverloads constructor(
...
     val frameCount = (durationMillis * fps) / 1000 + 1
     val startNanos = TimeUnit.MILLISECONDS.toNanos(start)
-    takeSnapshots(view, name, deviceConfig, theme, renderingMode, startNanos, fps, frameCount)
+    takeSnapshots(view, name, startNanos, fps, frameCount)
   }
 
-  private fun takeSnapshots(
-    view: View,
-    name: String?,
+  fun unsafeConfig(
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:220]: nit: updateUnsafeConfig?

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt [hunk 2]
```diff
@@ -229,50 +214,55 @@ class Paparazzi @JvmOverloads constructor(
...
-              // Required to reset underlying parser stream
-              layoutPullParser = LayoutPullParser.createFromString(contentRoot)
-          )
-
-      if (deviceConfig != null) {
-        sessionParamsBuilder = sessionParamsBuilder.copy(deviceConfig = deviceConfig)
-      }
+    require(deviceConfig != null || theme != null || renderingMode != null) {
+      "Calling updateConfig requires at least one non-null argument."
+    }
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:227]: nice

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt [hunk 3]
```diff
@@ -198,27 +192,18 @@ class Paparazzi @JvmOverloads constructor(
...
-  fun snapshot(
-    view: View,
-    name: String? = null,
-    deviceConfig: DeviceConfig? = null,
-    theme: String? = null,
-    renderingMode: RenderingMode? = null
-  ) {
-    takeSnapshots(view, name, deviceConfig, theme, renderingMode, 0, -1, 1)
+  fun snapshot(view: View, name: String? = null) {
+    takeSnapshots(view, name,0, -1, 1)
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:200]:
```suggestion
    takeSnapshots(view, name, 0, -1, 1)
```

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt [hunk 4]
```diff
@@ -183,13 +183,7 @@ class Paparazzi @JvmOverloads constructor(
 
   fun <V : View> inflate(@LayoutRes layoutId: Int): V = layoutInflater.inflate(layoutId, null) as V
 
-  fun snapshot(
-    name: String? = null,
-    deviceConfig: DeviceConfig? = null,
-    theme: String? = null,
-    renderingMode: RenderingMode? = null,
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:190]: bye

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt [hunk 5]
```diff
@@ -198,27 +192,18 @@ class Paparazzi @JvmOverloads constructor(
...
+    snapshot(parent, name)
   }
 
   @JvmOverloads
-  fun snapshot(
-    view: View,
-    name: String? = null,
-    deviceConfig: DeviceConfig? = null,
-    theme: String? = null,
-    renderingMode: RenderingMode? = null
```
Comment (jrodbx) [inline paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:210]: bye

File: paparazzi-gradle-plugin/src/test/projects/verify-rendering-modes/src/test/java/app/cash/paparazzi/plugin/test/RenderingModesTest.kt [hunk 6]
```diff
@@ -32,9 +32,11 @@ class RenderingModesTest {
...
 
     paparazzi.snapshot(view = linearLayout) // defaults to NORMAL
-    paparazzi.snapshot(view = linearLayout, renderingMode = RenderingMode.H_SCROLL)
+    paparazzi.unsafeConfig(renderingMode = RenderingMode.H_SCROLL)
+    paparazzi.snapshot(view = linearLayout)
 
+    paparazzi.unsafeConfig(renderingMode = RenderingMode.V_SCROLL)
     linearLayout.orientation = LinearLayout.VERTICAL
-    paparazzi.snapshot(view = linearLayout, renderingMode = RenderingMode.V_SCROLL)
+    paparazzi.snapshot(view = linearLayout)
```
Comment (jrodbx) [inline paparazzi-gradle-plugin/src/test/projects/verify-rendering-modes/src/test/java/app/cash/paparazzi/plugin/test/RenderingModesTest.kt:40]: nice

File: paparazzi-gradle-plugin/src/test/java/app/cash/paparazzi/gradle/PaparazziPluginTest.kt [hunk 7]
```diff
@@ -871,6 +871,26 @@ class PaparazziPluginTest {
...
+    val snapshots = snapshotsDir.listFiles()?.sortedBy { it.lastModified() }
+    assertThat(snapshots!!).hasSize(2)
+
+    val pixel3SnapshotImage = snapshots[0]
+    val nexus7SnapshotImage = snapshots[1]
+    val pixel3GoldenImage = File(fixtureRoot, "src/test/resources/pixel_3_launch.png")
+    val nexus7GoldenImage = File(fixtureRoot, "src/test/resources/nexus_7_launch.png")
+    assertThat(pixel3SnapshotImage).isSimilarTo(pixel3GoldenImage).withDefaultThreshold()
+    assertThat(nexus7SnapshotImage).isSimilarTo(nexus7GoldenImage).withDefaultThreshold()
+  }
```
Comment (jrodbx) [inline paparazzi-gradle-plugin/src/test/java/app/cash/paparazzi/gradle/PaparazziPluginTest.kt:892]: awesome!

**Review by jrodbx — APPROVED** [review]

File: paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt (hunk 1 above)
Comment (luis-cortes) [reply paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:220]: How about `unsafeUpdateConfig`? The config itself isn't what's unsafe.

Comment (jrodbx) [reply paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt]: True.  I'm down.

**Pull request merged**